### DISCUSSION TITLE
Just wanted to say...

### DISCUSSION THREAD

**Discussion by dodexahedron** [tui-cs/Terminal.Gui#3280] [Show and tell]
It's been SO NICE to be able to perform re-bases without trouble, now that normal merges are being used!

That is all. 😅

**Comment by tig** [comment]
Sorry I was dum before. Thanks for helping me be less dum. ;-)

**Reply by dodexahedron** [reply]
Bah no worries.

Definitely one of those things that you don't realize how much you take it for granted until it ain't working how you expect.

Now that it's fixed, though, remember that you don't need to (and really kinda shouldn't) merge v2_develop back into your own branch before submitting a PR, any more (at least for anything that has been branched from or rebased on v2_develop since changing the merge style.

Instead, re-base on it. _If_ you have conflicts, they will typically be a lot easier to deal with, depending on how granular the commits are between your original branch point and v2_develop/HEAD.

**Reply by dodexahedron** [reply]
Also, rebase won't cause pseudo-duplicate commits like merge can, since merge implies bringing along the history of both between branch point and merge point. So if you merge v2_develop into your branch before a PR, there's a pretty decent chance you'll see at least some of those dupes.

Basically means rebase, for work on forks. Merge from fork to origin, for PRs.

**Reply by tig** [reply]
I'm gonna be perfectly honest. 

I have no clue how to use rebase correctly.

It would be helpful to me if you showed me the commands you use in a typical PR.

**Reply by BDisp** [reply]
I'm gonna be perfectly honest too. :-)
I only use VS UI for git operations and rarely use command line.
I rarely use the rebase because it's prone to loose some changes for ever if it isn't done well. Now I prefer use only the merge and only use the rebase when it's really need and force a push.
But I'm also curious to learn even more.

**Reply by dodexahedron** [reply]
First an explanation of what the difference is:

## Re-base
Rebase is for changing the point from which a branch is branched. It's basically exactly what you're intending to do, when you merge from the mainline branch back into your own branch, except that it's done in the proper historical order.

It turns back the clock and then re-plays the commits in the order they happened.

If you use most GUIs for it, you get a nice easy way to address each individual conflict that may have happened _in the actual context it happened_, which is a major benefit over merge, which just takes the whole thing and tries to do it all at once. If you do it on the command line, it'll pause for conflict resolution at each conflict, as well. But now you get to understand how, when, where, and why something conflicted, rather than having to try to guess which one is correct.

Re-base keeps your work on your branch and other work wherever it came from.

When your code is supposed to be working on top of/from a tag, branch, or commit, re-base is the proper thing to do.

## Merge
Merge takes all changes from incoming branch A as a single change and uses it as the next commit on the target branch B. It represents a branch joining another. In the typical case, that's a branch which split from master being finished and re-joining master.

That's great when a piece of work is done, and is the right thing to do, there.

## Results
The end result, in terms of the code, if you did each method correctly, is the same.

But what it means in terms of source control is very different and creates difficulties that are not quite as severe as the difficulties caused by things such as the way merges were being done until recently, but still broken and tend to be major regression-factories as well.

It creates diamond ancestry, for subsequent commits, and that's not restricted to just your branch, either. It's a permanent thing that happened in history and can only be un-done by changing history all the way back from the point it was done, even if that's 1200 commits across 60 PRs and 4 months later.

When's that even relevant? When someone is trying to find the source of a bug (that's what git blame does). When walking backward, those merge commits create a fork in the road and you have to go down both forks to find out where the change actually came from, _especially_ if any conflict resolution had been done during the merge commit, because those changes do not have a parent - they exist only in that commit. That makes it even harder to backtrack from that point.

If a rebase had been done, the purity of each branch would have been preserved, and the source branch for a given change would be clear. Merges, in this situation, have an Oedipus complex.

Re-base also conveys intent much more clearly. When looking at commit history, you should be able to assume that the base of a branch (the commit it branches from) is the point in history with the code that _that branch_ intends to fix/enhance/whatever. This isn't just a convenience thing. It's what causes the difficulty mentioned earlier. If you re-base instead of merging, you have changed the graph to show that the new base is what this branch builds on top of, and you have left other work alone.

If there were merges back into this branch from its parent at any point, you basically usurped mainline and made your branch the boss for the run of its commits plus your merge which, in addition to the difficulties above, also can lead to the merge back into its parent _not_ reporting conflicts that should have been there, but are now "hidden" as a simple change due to being just a normal incoming change for the merge. That's another regression factory.

Github desktop, the git windows GUI, and the VS git extension all give an ok experience for rebasing, though they are pretty simplistic. They definitely do get the job done, though.

I use GitKraken, which is just....super-nice. [There's a free version, too](https://www.gitkraken.com/download/windows64).

I use a team licensed version, but the free edition does have the visual re-base feature as well as most other relevant features. The limitations are usually just that it will only allow use on public repos for some of the fancier features. This is a public repo, so that's not an issue. :)

**Reply by dodexahedron** [reply]
Actually... Ya know what?

I'm about to do a rebase, so how about I show some screenshots to give a visual to associate with all of that?

Stand by...

**Reply by dodexahedron** [reply]
Ok...

So here's how things look now, after the latest PR merges.

 - v2_develop is the leftmost line. This is looking at my fork, but I don't touch v2_develop on my fork, so it's identical to origin.
 - The second line is v2_develop in @tig's fork
 - The third line is @tig's mouse event work branch in his fork
 - The fourth line is my branch that made csproj changes
 - The fifth line is v2_develop at origin, but from an earlier point in time, due to improper merge activity.
 - The rightmost line is my branch working on FindHotKey and related code, branched from the end of the fourth line, to separate work and history and continue work without waiting for PR merge.
![image](https://github.com/gui-cs/Terminal.Gui/assets/8242181/a3761e0d-a8c7-4032-998b-77168fd4b419)

I'll continue in the next comment.

**Reply by dodexahedron** [reply]
So... Now that some merges to mainline have been done since I started work, I want my newest work to be up-to-date with that.

So, this is a rebase.

Just for reference, here is the documentation on the `git rebase` operation at the cli: https://git-scm.com/docs/git-rebase

The desired result is for the rightmost line to now start from the end of v2_develop, so that it is on top of the most current code.

In GitKraken, this is a simple drag-and-drop operation, dragging my branch on top of v2_develop, and choosing "rebase."

At the command line, the procedure is:
 1. Check out/switch to the branch that you are working in that needs to be rebased.
 2. Now do the rebase: `git rebase --onto v2_develop`

Your branch will now be re-played, in order, on top of v2_develop.

If there are any conflicts at any of the commits, it will "fail" which is just pausing to let you handle the conflicts. If you just want to take the base version of the code and drop that commit from your branch, use `git rebase --skip` to skip that commit and continue the rebase. Otherwise, resolve the conflict and then use `git rebase --continue` to continue to rebase. Any changes you make during conflict resolution will now actually be part of the commit that was being processed at that time, meaning you no longer have any sort of "fixing merge conflict" commits necessary, any more.

Once it has finished, your local copy now looks something like this.

Notice that there's still the original sixth line showing up. That's the remote copy of the branch still on my fork in github. My local version of that branch is now on top of v2_develop - all 17 commits of it.

![image](https://github.com/gui-cs/Terminal.Gui/assets/8242181/d2eb3c62-08ee-49da-af13-67bd06e2aa58)

**Reply by dodexahedron** [reply]
Now all you have to do is push it.

However, you just re-wrote history (which is ok, within and isolated to your own branch in your own fork).

So, you must either force push, if you want to keep the same branch name as before, or you can just push to a new branch. Whatever you like. I usually force push because I don't need those other dead-end branches hanging around.

So, push whichever way you like, and now your branch is up to date with mainline, has all your changes, and has conflicts resolved. It is now ready for a PR or continued work, if you're not done. In this example, I'm continuing work from the rebased branch, since I'm not done but wanted to be working from the latest code.

After the force push, now we have:

![image](https://github.com/gui-cs/Terminal.Gui/assets/8242181/b4bec501-4a31-46ee-8e0c-d347a972f4a6)

As you can see, now there are only 5 lines, because what used to be a sixth line due to being a branch from another branch is now just a line continuing straight up the top, because that branch is now a branch from v2_develop - its new base.

And then life carries on like nothing happened.

**Reply by dodexahedron** [reply]
The two commits that happened before the last one on v2_develop, and also the commit where the blue one comes into the purple one would have never happened or been necessary, had they been rebases. 

If I'm trying to walk back and find where a change that broke something came from, I have to split my search in 3 places to resolve the ambiguities created by those merges (not counting the legit merges to origin/v2_develop).

**Reply by dodexahedron** [reply]
Also, the more specific your individual commits are, the more powerful and simultaneously _easier_ this becomes, which is a rare combo. I didn't have to resolve any conflicts on this rebase nor on most of the others I've performed since the merge strategy was corrected.

Had my commits been bigger work dumps, I might have had a couple of conflicts to fix. But you fix them once and then they're fixed as if there never were a conflict. This is just one of many reasons I make so many commits. :)

**Reply by dodexahedron** [reply]
This all works the way it does because git really is basically just a transaction log.

Any given commit is NOT actually a full picture of the state of that branch at that point in time. It is the deltas, in order, from its base ref, to that point. So, rebasing is a fairly trivial operation, because it's just replaying those deltas. As long as that delta can be reconciled with the new base version, it all happens automatically with no fuss and no interaction.

**Reply by dodexahedron** [reply]
All that being said...

If you have a branch which you have already done some "bad" merging on, you shouldn't try to rebase it on any branch that you merged into it, unless you first reset to a point before that merge. There are a few possible results, none of which are great. Most often, it'll simply fail due to unresolvable history. The rest of the time, it'll end up duplicating commits _from the other branches_, along with all sorts of other unpleasant stuff. And that's because you now inextricably tied that branch's history to whatever happened between the original base and the commit you're merging from the source into your branch, but in a different order than it really happened. That could be hundreds of commits, covering several PRs. But in no case is it a good thing.

So just don't do it. :)

**Reply by dodexahedron** [reply]
There's also the `--interactive` option for the `git rebase` command, which I occasionally use.

What that does is, before it replays the commits on top of the new base, it gives you an opportunity to squash, reword, re-order, or drop any commits you've made. Play with that option if you like, but be careful until you're comfortable with things, because that whole log nature of git can make the impact of things like re-ordering commits non-obvious. Using a GUI tool for that is a good idea. GitKraken does it quite nicely, but I don't know if the free version has that feature enabled or not.

**Comment by dodexahedron** [comment]
Ha... The git discussion sub-thread in here points out a potential idea for a Terminal.Gui application: A TUI for git.
